src/coreclr/vm/jitinterface.cpp

```diff
@@ -7239,8 +7239,34 @@ bool getILIntrinsicImplementationForInterlocked(MethodDesc * ftn,
     if (ftn->GetMemberDef() != CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_T)->GetMemberDef())
         return false;
 
-    // Get MethodDesc for non-generic System.Threading.Interlocked.CompareExchange()
-    MethodDesc* cmpxchgObject = CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_OBJECT);
+    // Determine the type of the generic T method parameter
+    _ASSERTE(ftn->HasMethodInstantiation());
+    _ASSERTE(ftn->GetNumGenericMethodArgs() == 1);
+    TypeHandle typeHandle = ftn->GetMethodInstantiation()[0];
+    MethodTable* methodTable = typeHandle.GetMethodTable();
+
+    MethodDesc* cmpxchg;
+
+    // Based on the generic method parameter, determine which overload of CompareExchange
+    // to delegate to, or if we can't handle the type at all.
+    if (!typeHandle.IsValueType())
+    {
+        cmpxchg = CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_OBJECT);
+    }
+    else if (CorTypeInfo::IsPrimitiveType(typeHandle.GetVerifierCorElementType()))
+    {
+        UINT size = typeHandle.GetSize();
+        _ASSERTE(size == 1 || size == 2 || size == 4 || size == 8);
+        cmpxchg =
+            size == 1 ? CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_BYTE) :
+            size == 2 ? CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_USHRT) :
+            size == 4 ? CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_UINT) :
+            CoreLibBinder::GetMethod(METHOD__INTERLOCKED__COMPARE_EXCHANGE_ULONG);
+    }
+    else
+    {
+        return false;
+    }
 
     // Setup up the body of the method
     static BYTE il[] = {
```

Comment:
This is static memory. You need to have arrays of these so that each primitive type gets its own copy.